### ISSUE TITLE
Unable to update cluster from 1.20 -> 1.21, error running task "BootstrapClientTask/BootstrapClient" failed to verify token (received status code 403 from STS)

### ISSUE THREAD

**Issue by salavessa** [kubernetes/kops#12299]
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
1.21.1

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.21.4

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
Update existing cluster kops 1.20.2->1.21.1 and k8s 1.20.9->1.21.4.

**5. What happened after the commands executed?**
The worker nodes never join the cluster (the control-plane started and runs with no issues).
The `kops-configuration.service` on every node ends up failing after being unable to complete the below:
```
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: I0910 18:10:54.522051    1371 service.go:360] Enabling service "systemd-timesyncd"
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: W0910 18:10:54.689371    1371 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (9m30s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: I0910 18:10:54.689404    1371 executor.go:111] Tasks: 73 done / 81 total; 1 can run
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: I0910 18:10:54.689426    1371 executor.go:186] Executing task "BootstrapClientTask/BootstrapClient": BootstrapClientTask
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: W0910 18:10:54.963891    1371 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (9m29s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
Sep 10 18:10:54 ip-172-17-8-239 nodeup[1371]: I0910 18:10:54.963917    1371 executor.go:155] No progress made, sleeping before retrying 1 task(s)
Sep 10 18:11:04 ip-172-17-8-239 nodeup[1371]: I0910 18:11:04.965107    1371 executor.go:111] Tasks: 73 done / 81 total; 1 can run
Sep 10 18:11:04 ip-172-17-8-239 nodeup[1371]: I0910 18:11:04.965157    1371 executor.go:186] Executing task "BootstrapClientTask/BootstrapClient": BootstrapClientTask
Sep 10 18:11:05 ip-172-17-8-239 nodeup[1371]: W0910 18:11:05.268361    1371 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (9m19s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
Sep 10 18:11:05 ip-172-17-8-239 nodeup[1371]: I0910 18:11:05.268387    1371 executor.go:155] No progress made, sleeping before retrying 1 task(s)
...
```

**6. What did you expect to happen?**
New nodes to join the cluster.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2019-11-19T15:28:59Z"
  generation: 23
  name: name.cluster.com
spec:
  additionalPolicies:
    master: '[{"Effect":"Allow","Action":["sts:AssumeRole"],"Resource":["arn:aws:iam::000000000000:role/elasticsearch_logs"]}]'
    node: '[{"Effect":"Allow","Action":["sts:AssumeRole"],"Resource":["arn:aws:iam::000000000000:role/elasticsearch_logs"]}]'
  additionalSans:
  - name.cluster.com.othername.com
  api:
    dns: {}
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://state-store/name.cluster.com
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2c
      name: c
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-us-west-2c
      name: c
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - 10.119.1.0/24
  kubernetesVersion: 1.21.4
  masterInternalName: api.internal.name.cluster.com
  masterPublicName: api.name.cluster.com
  networkCIDR: 172.16.0.0/22
  networkID: vpc-00000000000000000
  networking:
    flannel:
      backend: vxlan
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 10.119.1.0/24
  subnets:
  - cidr: 172.16.0.0/27
    egress: nat-00000000000000000
    name: us-west-2a
    type: Private
    zone: us-west-2a
  - cidr: 172.16.0.32/27
    egress: nat-11111111111111111
    name: us-west-2b
    type: Private
    zone: us-west-2b
  - cidr: 172.16.0.64/27
    egress: nat-22222222222222222
    name: us-west-2c
    type: Private
    zone: us-west-2c
  - cidr: 172.16.0.96/28
    name: utility-us-west-2a
    type: Utility
    zone: us-west-2a
  - cidr: 172.16.0.112/28
    name: utility-us-west-2b
    type: Utility
    zone: us-west-2b
  - cidr: 172.16.0.128/28
    name: utility-us-west-2c
    type: Utility
    zone: us-west-2c
  topology:
    dns:
      type: Private
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-11-19T15:28:59Z"
  generation: 17
  labels:
    kops.k8s.io/cluster: name.cluster.com
  name: master-us-west-2c
spec:
  additionalSecurityGroups:
  - sg-00000000000000000
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210907
  machineType: t3a.medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-west-2c
  role: Master
  rootVolumeEncryption: true
  rootVolumeOptimization: true
  rootVolumeSize: 32
  subnets:
  - us-west-2c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2019-11-19T15:29:00Z"
  generation: 21
  labels:
    kops.k8s.io/cluster: name.cluster.com
  name: nodes
spec:
  additionalSecurityGroups:
  - sg-00000000000000000
  - sg-11111111111111111
  detailedInstanceMonitoring: true
  externalLoadBalancers:
  - targetGroupArn: arn:aws:elasticloadbalancing:us-west-2:000000000000:targetgroup/tg1/0000000000000000
  - targetGroupArn: arn:aws:elasticloadbalancing:us-west-2:000000000000:targetgroup/tg2/1111111111111111
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210907
  machineType: r5.2xlarge
  maxSize: 3
  minSize: 3
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  rootVolumeEncryption: true
  rootVolumeOptimization: true
  rootVolumeSize: 100
  subnets:
  - us-west-2a
  - us-west-2b
  - us-west-2c
  suspendProcesses:
  - AZRebalance
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
n/a

**9. Anything else do we need to know?**
After manually updating the node's "Launch template" to use the previous existing version (created with kops 1.20.2 / k8s 1.20.9) instead of "Latest", the newly created nodes (running k8s 1.20.9) are still able to join the cluster.

**Comment by olemarkus** [comment]
This may happen if you roll nodes before the masters (or should CAS spin up a new node before all masters have rolled)

The reason for this is the switch to use regional STS endpoints. See https://github.com/kubernetes/kops/pull/12043 and associated issue.

**Comment by olemarkus** [comment]
You can do `kops rolling-update cluster --instance-group-roles master --yes --force` to force roll. This shouldn't really be necessary as the kops-controller change should mark nodes with `NeedsUpdate`. But perhaps there is a bug that prevents nodes from being marked.

Are the clusters single-master in both your cases?

**Comment by salavessa** [comment]
@caiconkhicon, if @olemarkus command doesn't do it, just try terminating the master instance(s) directly from AWS EC2 console or AWS CLI - the ASG will bring in a new master node and after the new master joins the cluster, updated nodes should be able to join the cluster too.

Some further updates after updating more than 15 kubernetes clusters: this happened to me in every k8s cluster which had a single master node and the networking was flannel-vxlan.
Multi-master clusters running flannel-vxlan and single-master not using flannel (using kubenet) were not affected.

**Comment by caiconkhicon** [comment]
Thank @olemarkus and @salavessa 
In my case, my cluster has 3 master nodes and running flannel-vxlan, and I am updating from 1.19 to 1.21. 
I will try your solutions, and if they don't work, I think I will consider udate to 1.20 only.

**Comment by olemarkus** [comment]
I suspect that you also run rolling update quite quickly after kops update?
Working theory I think the addon update happens after the new master has started, but before the master is labelled.
If this is the case, it probably isn't that easy to resolve from kops side.

To mitigate this, I would pause for 60s after the kops update.

**Comment by salavessa** [comment]
Hi @olemarkus, I actually always watch all pods running in the cluster (`kubectl get pod -Aw`) after the `kops update cluster --yes` at least for a couple of minutes and wait for any pods to be recreated (usually dns and flannel) before starting the rollout.

As a side note and maybe not related with this specific issue but I think it's worth mentioning that there was a change in `kops 1.21` which broke the `kube-flannel-ds` daemonset update as the `selector` immutable field got updated: `role.kubernetes.io/networking: "1"` is no longer present - [commit b44065c, file `upup/models/cloudup/resources/addons/networking.flannel/k8s-1.12.yaml.template`](https://github.com/kubernetes/kops/commit/b44065c6927ec48c7849363a4b5e2a26e01d79a5#diff-0afa1bc6d9dc8cc7f9834372e9ac45ded576cde6eaf18b3c28ded7bcd22c24d9L160).

The fix/workaround is actually pretty simple as one just needs to delete the flannel daemonset itself (`kubectl --namespace=kube-system delete daemonsets.apps kube-flannel-ds`) and it will get recreated in less than a minute by `protokube`.

**Comment by caiconkhicon** [comment]
@olemarkus  @salavessa : I recreated all 3 master nodes and recreated one worker node. The issue persists: 
```
kops-configuration.service - Run kOps bootstrap (nodeup)
     Loaded: loaded (/lib/systemd/system/kops-configuration.service; disabled; vendor preset: enabled)
     Active: activating (start) since Tue 2021-09-21 20:16:53 UTC; 2min 14s ago
       Docs: https://github.com/kubernetes/kops
   Main PID: 1388 (nodeup)
      Tasks: 17 (limit: 18645)
     Memory: 1.2G
     CGroup: /system.slice/kops-configuration.service
             └─1388 /opt/kops/bin/nodeup --conf=/opt/kops/conf/kube_env.yaml --v=8

Sep 21 20:18:45 ip-172-38-16-9 nodeup[1388]: W0921 20:18:45.258772    1388 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (8m15s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorR>
Sep 21 20:18:45 ip-172-38-16-9 nodeup[1388]: I0921 20:18:45.258798    1388 executor.go:155] No progress made, sleeping before retrying 1 task(s)
Sep 21 20:18:55 ip-172-38-16-9 nodeup[1388]: I0921 20:18:55.260935    1388 executor.go:111] Tasks: 71 done / 79 total; 1 can run
Sep 21 20:18:55 ip-172-38-16-9 nodeup[1388]: I0921 20:18:55.260981    1388 executor.go:186] Executing task "BootstrapClientTask/BootstrapClient": BootstrapClientTask
Sep 21 20:18:55 ip-172-38-16-9 nodeup[1388]: W0921 20:18:55.626348    1388 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (8m5s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorRe>
Sep 21 20:18:55 ip-172-38-16-9 nodeup[1388]: I0921 20:18:55.626372    1388 executor.go:155] No progress made, sleeping before retrying 1 task(s)
Sep 21 20:19:05 ip-172-38-16-9 nodeup[1388]: I0921 20:19:05.628692    1388 executor.go:111] Tasks: 71 done / 79 total; 1 can run
Sep 21 20:19:05 ip-172-38-16-9 nodeup[1388]: I0921 20:19:05.628752    1388 executor.go:186] Executing task "BootstrapClientTask/BootstrapClient": BootstrapClientTask
Sep 21 20:19:05 ip-172-38-16-9 nodeup[1388]: W0921 20:19:05.998310    1388 executor.go:139] error running task "BootstrapClientTask/BootstrapClient" (7m55s remaining to succeed): bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorR>
Sep 21 20:19:05 ip-172-38-16-9 nodeup[1388]: I0921 20:19:05.998336    1388 executor.go:155] No progress made, sleeping before retrying 1 task(s)
```

Does any of you have any other ideas how to solve this?

On a side node, I also tried creating a node using the old Launch Template - which used k8s 1.19. This node could join the cluter, but was never Ready, because of this error with the `kube-flannel` pod:

```
  Warning  FailedCreatePodSandBox  3m30s (x187 over 43m)  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to generate seccomp spec opts: seccomp is not supported
```
Is this the one you mentioned @salavessa ?

Thank you guys a lot.

**Comment by caiconkhicon** [comment]
And does it have anything to do with the change from gp2 -> gp3 and docker -> containerd between kops 1.19 and 1.21?

**Comment by salavessa** [comment]
@caiconkhicon I didn't experience any issues updating from `kops 1.19 (k8s 1.19)` to `kops 1.20 (k8s 1.20)`, and it was during that update that I also moved from `docker` to `containerd` and `gp2` to `gp3` (maybe you can try this gradual update instead of going from 1.19 to 1.21).

The only issues I got are the ones mentioned in this issue when updating `kops 1.20 (k8s 1.20)` to `kops 1.21 (k8s 1.21)`:
1. nodes running k8s 1.21 unable to join the cluster after the master was rolled-out for the first time
2. flannel daemonset definition not updated: which maybe wasn't actually a "real" issue updating from 1.20 to 1.21 because the selector is the only thing that changed here (unless there are some bootstrap processes that don't complete because the `kubectl update` triggered by `protokube` command keeps failing).
I still believe this should be sorted somehow (even if it's just by documenting that flannel daemonset must be manually deleted) as it will certainly block future flannel daemonset updates for clusters created with `kops <=1.20` as the daemonset definition can't be "patched" due to the immutable selector field update.

_NOTE 1_: @olemarkus, another thing I tried was to reboot (instead of terminating) the updated master node (right after first rollout) but this didn't seem to do the trick as new updated nodes were still not able to join the cluster. Terminating the master instance seemed to be the only way to trigger whatever made it work again and, just guessing here: maybe there's some ETCD configuration/entry that only gets refreshed when an updated master gets "refreshed" for the second time?

_NOTE 2_: While I was experiencing the issues I also created a cluster from scratch using `kops 1.22` with the exact same specs and everything started just fine, so it seems to be related with the cluster update only.

**Comment by olemarkus** [comment]
Sorry, but this is getting a bit unwieldy. There are different issues being grouped together into this one ticket that is already closed. Can you please open up individual issues with the details provided here? Especially with steps that makes it possible for us to reproduce

**Comment by salavessa** [comment]
@olemarkus I'm reopening this issue which I closed because the [workaround ](#issuecomment-917374799) (second rollout for master nodes) worked for me but the underlying issue seems to be real and unknown (maybe _NOTE 1_ in [my previous comment](#issuecomment-924732834) can provide extra clues)?

The key points seem to be: `single master` + `flannel-vxlan` + `kops update 1.20->1.21` = `master must be rolled-out twice before new nodes are able to join cluster`


A separate issue #12388 relative to `kube-flannel-ds` daemonset update issue mentioned [here](#issuecomment-924040060) and [here](#issuecomment-924732834) has been created.

**Comment by caiconkhicon** [comment]
Another test from me: I use Kops 1.21, then deploy a K8S 1.19 cluster, then update it to 1.21. It works without any problem.

**Comment by caiconkhicon** [comment]
From the issue that causes the PR that @olemarkus mentioned #12043, it seems that it is not fixed completely yet? https://github.com/kubernetes/kops/issues/12042#issuecomment-891469956

**Comment by caiconkhicon** [comment]
For whoever faces this issue and has the disaster of cannot operate a production cluster: Downgrading Kops to 1.20.2 (Kops, not only K8S) fixes the issue. You can use this command: (after downloading Kops 1.20.2)
```
kops update cluster --allow-kops-downgrade --yes
```
then recreate the new worker nodes using the new Launch Template.

And for others: Please DO NOT use Kops 1.21, if you don't want to face this issue like me.

**Comment by olemarkus** [comment]
This is not very good advice. There is some specific configurations that are causing this and that has proven tricky to pin down. There is way too little activity here or in slack to suggest there is a general problem with 1.21.

I think the only known issue we have found here is if someone has explicitly disabled regional STS on their account. Regional STS endpoints is what was introduced in #12043

**Comment by caiconkhicon** [comment]
Hi @olemarkus , I agree with you that it is too much to tell people not to use Kops 1.21, however it is not a bad choice if someone wants to stay on the safe side. 

Regarding STS, I checked and my account uses the default configure, so enable in all region except some special ones (which I don't use). Can it be some IAM permission issue?

**Comment by salavessa** [comment]
Hi @olemarkus I can also confirm that all STS regional endopints are (and have always been) active in my AWS accounts - actually that was one of the things I checked before opening this issue as it was the main thing the error was suggesting to me.

Moreover, I faced this issue in multiple clusters which are hosted along **three different AWS accounts** and, for each account, I have clusters in a **couple of different regions** (Ohio and Oregon).
All of them were operational again after the [master's second rollout workaround](#issuecomment-917374799).

**Comment by olemarkus** [comment]
The master second roll is something that is known. We have other bugs related to that as well, but this is only an issue on single-master clusters. And I believe that is your case as well @salavessa. 

IAM is not involved in this, but bootstrapping nodes and new kops-controller need to both use the regional endpoints. Those 403 suggests that is not the case. I.e kops-controller must run image version 1.21.1. @salavessa if you can reproduce this one more time, maybe you can check the kops-controller image version.

**Comment by salavessa** [comment]
@olemarkus that's correct, this issue only happened in `single-master` + `flannel-vxlan` clusters (private topology).
I also had some `kubenet` clusters (public topology) but all of them updated with no issues.

Unfortunately I have rolled out all of my  `single-master` + `flannel-vxlan` clusters already so won't be able to check  kops-controller image version for you, sorry.

**Comment by pascal-hofmann** [comment]
This issue still persists. We currently have a 3 master cluster where we can't bring up new nodes.

- kops-controller image is version 1.21.4 
- regional and global STS endpoint is active in the account
- Error message: `bootstrap returned status code 403: failed to verify token: received status code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">`
- We tried the forces re-roll of masters - unfortunately without success. :(
- We will now try terminating one master. If this does not work I have no further ideas.

**Comment by pascal-hofmann** [comment]
We found the issue. Due to broken automation we managed to have terraform generated by kops 1.21.0 applied while rolling updates with kops 1.21.4 🤦.

A clean and proper update to 1.21.4 fixed all issues for us.
